src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/TypeMapLazyDictionary.cs

```diff
@@ -0,0 +1,310 @@
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+using System.Reflection.Metadata;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace System.Runtime.InteropServices
+{
+    [RequiresUnreferencedCode("Lazy TypeMap isn't supported for Trimmer scenarios")]
+    internal sealed partial class TypeMapLazyDictionary
+    {
+        private readonly LazyExternalTypeDictionary _externalTypeMap;
+        private readonly LazyProxyTypeDictionary _proxyTypeMap;
+
+        private ref struct CallbackContext
+        {
+            [RequiresUnreferencedCode("Lazy TypeMap isn't supported for Trimmer scenarios")]
+            public CallbackContext()
+            {
+                ExternalTypeMap = new LazyExternalTypeDictionary();
+                ProxyTypeMap = new LazyProxyTypeDictionary();
+            }
+
+            public LazyExternalTypeDictionary ExternalTypeMap;
+            public LazyProxyTypeDictionary ProxyTypeMap;
+        }
+
+        // See assemblynative.hpp for native version.
+        public unsafe struct ProcessAttributesCallbackArg
+        {
+            public void* Utf8String1;
+            public void* Utf8String2;
+            public void* Utf8String3;
+            public int StringLen1;
+            public int StringLen2;
+            public int StringLen3;
+        }
+
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "TypeMapLazyDictionary_ProcessAttributes")]
+        private static unsafe partial void ProcessAttributes(
+            QCallAssembly assembly,
+            QCallTypeHandle groupType,
+            delegate* unmanaged<CallbackContext*, ProcessAttributesCallbackArg*, void> newExternalTypeEntry,
+            delegate* unmanaged<CallbackContext*, ProcessAttributesCallbackArg*, void> newProxyTypeEntry,
+            CallbackContext* context);
+
+        private static TypeName CreateNewTypeName(string typeName, string assemblyName)
+        {
+            AssemblyNameInfo asmNameInfo = AssemblyNameInfo.Parse(assemblyName);
+            TypeName parsedType = new TypeName(typeName, asmNameInfo);
+            Debug.Assert(parsedType.AssemblyName != null);
+            return parsedType;
+        }
+
+        [UnmanagedCallersOnly]
+        private static unsafe void NewExternalTypeEntry(CallbackContext* context, ProcessAttributesCallbackArg* arg)
+        {
+            Debug.Assert(context != null);
+            Debug.Assert(arg != null);
+            Debug.Assert(arg->Utf8String1 != null);
+            Debug.Assert(arg->Utf8String2 != null);
+
+            string externalTypeName = new((sbyte*)arg->Utf8String1, 0, arg->StringLen1, Encoding.UTF8);
+            string targetType = new((sbyte*)arg->Utf8String2, 0, arg->StringLen2, Encoding.UTF8);
+
+            TypeName parsedTarget = TypeNameParser.Parse(targetType, throwOnError: true)!;
```

Comment:
Any exceptions thrown here are going to turn into fail fasts at UnmanagedCallersOnly boundary (there is no exception interop on Unix). Do we want to propagate the exception all the way to the caller?

Reply:
These exceptions shouldn't ever happen. All these types are coming from actual metadata from a .NET compiler. If they are invalid, then the assembly is invalid. I can comment on that if you'd like?

Reply:
I agree that these exceptions should not happen in well-formed apps. I am just pointing out that this does not follow our typical exception handling pattern to propagate exceptions to the caller.

The fix would be to add try/catch here, add Exception to the context structure, and rethrow the exception from context in the main entrypoint. If you think that it is not worth it, comment would be nice.

Reply:
I'm wrong here. We need to handle this case as there are other scenarios that need exceptions to flow. I'll use `ExceptionDispatchInfo`.